presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/GroupInnerJoinsByConnectorRuleSet.java

```diff
@@ -0,0 +1,726 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static com.facebook.presto.SystemSessionProperties.INEQUALITY_JOIN_PUSHDOWN_ENABLED;
+import static com.facebook.presto.SystemSessionProperties.isInnerJoinPushdownEnabled;
+import static com.facebook.presto.common.function.OperatorType.GREATER_THAN;
+import static com.facebook.presto.common.function.OperatorType.GREATER_THAN_OR_EQUAL;
+import static com.facebook.presto.common.function.OperatorType.LESS_THAN;
+import static com.facebook.presto.common.function.OperatorType.LESS_THAN_OR_EQUAL;
+import static com.facebook.presto.common.function.OperatorType.NOT_EQUAL;
+import static com.facebook.presto.expressions.LogicalRowExpressions.TRUE_CONSTANT;
+import static com.facebook.presto.expressions.LogicalRowExpressions.and;
+import static com.facebook.presto.expressions.LogicalRowExpressions.extractConjuncts;
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.spi.connector.ConnectorCapabilities.SUPPORTS_JOIN_PUSHDOWN;
+import static com.facebook.presto.spi.plan.JoinType.INNER;
+import static com.facebook.presto.sql.planner.PlannerUtils.restrictOutput;
+import static com.facebook.presto.sql.planner.VariablesExtractor.extractUnique;
+import static com.facebook.presto.sql.planner.optimizations.JoinNodeUtils.toRowExpression;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.filter;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This optimizer attempts to group TableScanNode's of an inner-join graph that belong to the same connector
+ * This allows those connectors that can participate in join pushdown, to rewrite these sources to a new TableScanNode that represents the result of the pushed down join
+ * This re-written join graph has filter's pulled up, so filters need to be re pushed down again with the PredicatePushdown rule
+ * <p>
+ * Example 1:
+ * Before Transformation:
+ * --OutputNode
+ * `-- InnerJoin1
+ * |-- InnerJoin2
+ * |   |-- TableScanNode1
+ * |   |   `-- TableHandle1
+ * |   `-- TableScanNode2
+ * |       `-- TableHandle2
+ * `-- TableScanNode3
+ * `-- TableHandle3
+ * <p>
+ * Suppose that TableScanNode1, TableScanNode2 and TableScanNode3 are from the same catalog.
+ * <p>
+ * After Transformation:
+ * --OutputNode
+ * `--FilterNode(pulled up equi-join clause + join filters)
```

Comment:
The transformation looks like its a FilterNode -> TableScan Node.

Where is the filter pushdown into the tablescan expected to happen?

Reply:
The filters that will be pulled up iff they can later be pushed down by the connector in question; e.g the [JDBC filter pushdown rule](https://github.com/prestodb/presto/blob/9d0e329ead6faae8c466ecc8b2319a64cab67e1f/presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/optimization/JdbcComputePushdown.java) will do this for JDBC based connectors